Add `{From,To}HttpApiData` instances for `Iso8601Range`.

jonathanknowles opened the pull request:
# Issue Number

#466 

# Overview

I have:

- [x] Added `{From,To}Text` instances for `Iso8601Range`.
- [x] Added roundtrip tests for `{From,To}Text` instances for `Iso8601Range`.
- [x] Added set of negative tests for `{From,To}Text` instances for `Iso8601Range`.
- [x] Added `{From,To}HttpApiData` instances for `Iso8601Range`.
- [x] Added roundtrip tests for `HttpApiData` instances for `Iso8601Range`.

jonathanknowles committed:
Add `ToText` and `FromText` instances for `Iso8601Range`.

jonathanknowles committed:
Add `{From,To}HttpApiData` instances for `Iso8601Range` type.

jonathanknowles committed:
Add `Arbitrary` instance for `Iso8601Range` type.

jonathanknowles committed:
Add textual encoding roundtrip test for `Iso8601Range` type.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -546,6 +552,15 @@ instance Arbitrary ApiFee where
+        (mt1, mt2) <- arbitrary
+        let (mv1, mv2) =
+                if all isJust [mt1, mt2]
+                then (min mt1 mt2, max mt1 mt2)
```
Not sure to get this one? This seems to be forbidding `mt1` to be greater than `mt2` :thinking: ? (whereas that's a valid case for the range..)

jonathanknowles replied:
> Not sure to get this one? This seems to be forbidding `mt1` to be greater than `mt2` thinking ? (whereas that's a valid case for the range..)

Hmm. I don't see how that's a valid case.

From my understanding of `Iso8601Range v1 v2`:

* `v1` is an optional _earliest time_.
* `v2` is an optional _latest time_.

So this `Arbitrary` instance is designed to only generate the following cases:

* `Iso8601Range Nothing Nothing`
* `Iso8601Range Nothing (Just t)`
* `Iso8601Range (Just t) Nothing`
* `Iso8601Range (Just t1) (Just t2)` where `(t1 <= t2)`.

Perhaps I have misunderstood?

jonathanknowles replied:
> Not sure to get this one? This seems to be forbidding `mt1` to be greater than `mt2` thinking ? (whereas that's a valid case for the range..)

I've just read your later comment. I'll update the code to support both ascending and descending ranges.

jonathanknowles replied:
@KtorZ Now updated to allow ranges in both ascending and descending order.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -239,13 +248,65 @@ data ApiErrorCode
+                parseTimeM False defaultTimeLocale basicUtc $ T.unpack timeText
+
+        validRange :: Maybe UTCTime -> Maybe UTCTime -> Bool
+        validRange mt1 mt2 = fromMaybe True $ (<) <$> mt1 <*> mt2
```
Ah! Found it! 
This is _false_, the order between `mt1` and `mt2` conveys the ordering (ASC vs DESC) of the range, cf:

- `20190227T000000Z-20200227T000000Z`: means all transaction between 2019-02-27 and 2020-02-27, in ascending order.

- `20200227T000000Z-20190227T000000Z`: means all transaction between 2020-02-27 and 2019-02-27, in descending order.

jonathanknowles replied:
> Ah! Found it!
> This is _false_, the order between `mt1` and `mt2` conveys the ordering (ASC vs DESC) of the range, cf:
> 
>     * `20190227T000000Z-20200227T000000Z`: means all transaction between 2019-02-27 and 2020-02-27, in ascending order.
> 
>     * `20200227T000000Z-20190227T000000Z`: means all transaction between 2020-02-27 and 2019-02-27, in descending order.

Okay, that makes sense. I'll update the code to allow both ascending and descending ranges.

jonathanknowles replied:
@KtorZ Now updated to allow ranges in both ascending and descending order.

jonathanknowles committed:
Allow `Iso8601Range` to encode ranges that are either ascending or descending.

A range in *ascending*  order is `Iso8601Range t1 t2` for which `t1 < t2`.
A range in *descending* order is `Iso8601Range t1 t2` for which `t1 > t2`.

jonathanknowles committed:
Use `Data.Time.Text` to provide conversion functions between time values and text.

jonathanknowles committed:
Improve the Haddock comment for the `Iso8601Range` type.

jonathanknowles committed:
Add negative tests for textual decoding of `Iso8601Range` values.

jonathanknowles committed:
Add `HttpApiData` roundtrip test for `Iso8601Range` data type.

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -220,32 +227,121 @@ data ApiErrorCode
+--
+--     Example textual representation:
+--
+--     > "name *-*"
```
very nice instructions describing how to use it :100:

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -220,32 +227,121 @@ data ApiErrorCode
+      where
+        expectedPrefix = T.pack $ symbolVal (Proxy @name)
+
+        guardB :: Bool -> String -> Either TextDecodingError ()
```
probably there are already functions for that

paweljakubas commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -220,32 +227,121 @@ data ApiErrorCode
+        guardB :: Bool -> String -> Either TextDecodingError ()
+        guardB b err = if b then Right () else raise err
+
+        guardM :: Maybe a -> String -> Either TextDecodingError a
```
looks like : `https://hackage.haskell.org/package/either-5.0.1.1/docs/Data-Either-Combinators.html#v:maybeToRight`

paweljakubas reviewed (approved):
Cool! Small remarks to maybe use `maybeToEither`
